main/init/buildgen/src/mill/main/buildgen/BuildGenBase.scala

```diff
@@ -2,68 +2,102 @@ package mill.main.buildgen
 
 import mill.main.buildgen.BuildGenUtil.{buildPackages, compactBuildTree, writeBuildObject}
 
-import scala.collection.immutable.SortedMap
+import scala.collection.immutable.{SortedMap, SortedSet}
 
-trait BuildGenBase[M, D] {
+/*
+TODO Can we just convert all generic type parameters to abstract type members?
+ See https://stackoverflow.com/a/1154727/5082913.
+ I think abstract type members are preferred in this case.
+ */
+trait BuildGenBase[M, D, I] {
   type C
-  def convertWriteOut(cfg: C, shared: BuildGenUtil.Config, input: Tree[Node[M]]): Unit = {
+  def convertWriteOut(cfg: C, shared: BuildGenUtil.BasicConfig, input: I): Unit = {
     val output = convert(input, cfg, shared)
     writeBuildObject(if (shared.merge.value) compactBuildTree(output) else output)
   }
 
+  /**
+   * A possibly optional module model which might be a parent directory of an actual module without its own sources.
+   */
+  type OM
+  extension (om: OM) def toOption(): Option[M]
```

Comment:
Can we just use `Option[M]` rather than defining this alias?

Reply:
I added this so I could refactor less code and get it to work at the time being. I can switch to `Option[M]` but more Maven and Gradle implementation code will be refactored and generalized.

Reply:
Let's go with that then